pkg/gadgets/bindsnoop/tracer/core/tracer.go

```diff
@@ -0,0 +1,327 @@
+// #include "./bpf/bindsnoop.h"
+// #include <arpa/inet.h>
+// #include <stdlib.h>
+//
+//char *ip_to_string(const struct bind_event *event) {
+//	socklen_t size;
+//	int ip_type;
+//	char *ip;
+//
+//	if (event->ver == 4)
+//		ip_type = AF_INET;
+//	else if (event->ver == 6)
+//		ip_type = AF_INET6;
+//	else
+//		return NULL;
+//
+//	size = sizeof *ip * INET6_ADDRSTRLEN;
+//	ip = malloc(size);
+//  if (ip == NULL)
+//		return NULL;
+//
+//	inet_ntop(ip_type, &event->addr, ip, size);
+//
+//	return ip;
+//}
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+	"github.com/cilium/ebpf/perf"
+	containercollection "github.com/kinvolk/inspektor-gadget/pkg/container-collection"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/bindsnoop/tracer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/bindsnoop/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+	"github.com/vishvananda/netlink"
+)
+
+//go:generate sh -c "GOOS=$(go env GOHOSTOS) GOARCH=$(go env GOHOSTARCH) go run github.com/cilium/ebpf/cmd/bpf2go -target bpfel -cc clang bindsnoop ./bpf/bindsnoop.bpf.c -- -I./bpf/ -I../../../../ -target bpf -D__TARGET_ARCH_x86"
+
+type Tracer struct {
+	config        *tracer.Config
+	resolver      containercollection.ContainerResolver
+	eventCallback func(types.Event)
+	node          string
+
+	objs      bindsnoopObjects
+	ipv4Entry link.Link
+	ipv4Exit  link.Link
+	ipv6Entry link.Link
+	ipv6Exit  link.Link
+	reader    *perf.Reader
+}
+
+func NewTracer(config *tracer.Config, resolver containercollection.ContainerResolver,
+	eventCallback func(types.Event), node string) (*Tracer, error) {
+	t := &Tracer{
+		config:        config,
+		resolver:      resolver,
+		eventCallback: eventCallback,
+		node:          node,
+	}
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	return t, nil
+}
+
+func (t *Tracer) Stop() {
+	if t.ipv4Entry != nil {
+		t.ipv4Entry.Close()
+		t.ipv4Entry = nil
+	}
+	if t.ipv4Exit != nil {
+		t.ipv4Exit.Close()
+		t.ipv4Exit = nil
+	}
+
+	if t.ipv6Entry != nil {
+		t.ipv6Entry.Close()
+		t.ipv6Entry = nil
+	}
+	if t.ipv6Exit != nil {
+		t.ipv6Exit.Close()
+		t.ipv6Exit = nil
+	}
+
+	if t.reader != nil {
+		t.reader.Close()
+		t.reader = nil
+	}
+
+	t.objs.Close()
+}
+
+func (t *Tracer) start() error {
+	spec, err := loadBindsnoop()
+	if err != nil {
+		return fmt.Errorf("Failed to load ebpf program: %w", err)
+	}
+
+	filter_by_mnt_ns := false
+
+	if t.config.MountnsMap != "" {
+		filter_by_mnt_ns = true
+		m := spec.Maps["mount_ns_set"]
+		m.Pinning = ebpf.PinByName
+		m.Name = filepath.Base(t.config.MountnsMap)
+	}
+
+	filter_by_port := false
+	if len(t.config.TargetPorts) > 0 {
+		filter_by_port = true
+
+		m := spec.Maps["ports"]
+		m.Pinning = ebpf.PinByName
+		m.Name = "ports filtering map"
+
+		for _, port := range t.config.TargetPorts {
+			m.Contents = append(m.Contents, ebpf.MapKV{port, port})
```

Comment:
I am struggling with this snippet.
The goal is to fill the maps with ports given by the user so events will only be generated for this particular ports.
I think I am missing something with the `MapSpec`.